src/main.ts

```diff
@@ -1,256 +1,152 @@
-      return fab as any;
-    }
-    return (fab.isLoading() ? fab : this) as any;
-  }
-
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return this as any;
-  }
-
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onLoading;
-  }
+export interface Loading {
+  readonly _tag: 'Loading';
 }
 
-export class Failure<L, A> {
-  readonly _tag = 'Failure';
-  constructor(readonly value: L) {}
-
-  isNotAsked() {
-    return false;
-  }
-  isLoading() {
-    return false;
-  }
-  isFailure() {
-    return true;
-  }
-  isSuccess() {
-    return false;
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return this as any;
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    return (fab.isFailure() ? fab : this) as any;
-  }
-
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return this as any;
-  }
+export interface Failure<F> {
+  readonly _tag: 'Failure';
+  readonly failure: F;
+}
 
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onFailure(this.value);
-  }
+export interface Success<S> {
+  readonly _tag: 'Success';
+  readonly success: S;
 }
 
-export class Success<L, A> {
-  readonly _tag = 'Success';
-  constructor(readonly value: A) {}
+export type Dataway<F, S> = NotAsked | Loading | Failure<F> | Success<S>;
 
-  isNotAsked() {
-    return false;
-  }
-  isLoading() {
-    return false;
-  }
-  isFailure() {
-    return false;
-  }
-  isSuccess() {
-    return true;
-  }
-
-  map<B>(f: (a: A) => B): Dataway<L, B> {
-    return new Success(f(this.value));
-  }
-  ap<B>(fab: Dataway<L, (a: A) => B>): Dataway<L, B> {
-    if (fab.isSuccess()) {
-      const _fab = fab as Success<L, (a: A) => B>;
-      return this.map(_fab.value);
-    }
-    return fab as any;
-  }
+export const notAsked: Dataway<never, never> = { _tag: 'NotAsked' };
 
-  chain<B>(f: (a: A) => Dataway<L, B>): Dataway<L, B> {
-    return f(this.value);
-  }
+export const loading: Dataway<never, never> = { _tag: 'Loading' };
 
-  fold<B>(
-    onNotAsked: B,
-    onLoading: B,
-    onFailure: (e: L) => B,
-    onSuccess: (a: A) => B,
-  ) {
-    return onSuccess(this.value);
-  }
+export function failure<F = never, S = never>(f: F): Dataway<F, S> {
+  return { _tag: 'Failure', failure: f };
 }
 
-export const notAsked = <L, A>(): NotAsked<L, A> => {
-  return new NotAsked();
-};
-export const loading = <L, A>(): Loading<L, A> => {
-  return new Loading();
-};
-export const failure = <L, A>(l: L): Failure<L, A> => {
-  return new Failure(l);
-};
-export const success = <L, A>(a: A): Success<L, A> => {
-  return new Success(a);
-};
+export function success<F = never, S = never>(s: S): Dataway<F, S> {
+  return { _tag: 'Success', success: s };
+}
 
-const of = success;
+export function isNotAsked<L, A>(ma: Dataway<L, A>): ma is NotAsked {
+  return ma._tag === 'NotAsked' ? true : false;
```

Comment:
```suggestion
  return ma._tag === 'NotAsked';
```
non ?
et du coup pareil en dessous

Reply:
ahaha